Atomic/sign.py

```diff
@@ -0,0 +1,89 @@
+from . import util
+from . import Atomic
+import os
+import tempfile
+from .atomic import AtomicError
+from .client import ids_same, add_shaw
+
+
+class Sign(Atomic):
+    def sign(self):
+        # TODO
+        # If an alternate sign key is provided, should we verify that it is valid?
+        # Should we check to see if there are valid GPG signatures?
+        # Atomic is run as sudo. Should we work around that?
+
+        if self.args.debug:
+            util.write_out(str(self.args))
+
+        # Verify all the images provided are valid
+        self.check_input_validity()
+
+        signature_path = util.get_signature_path(self.atomic_config)
+
+        signer = util.get_atomic_config_item(['default_signer']) if \
+            self.args.sign_key is None else self.args.sign_key
+
+        # Make sure the signature path exists
+        if not os.path.exists(signature_path):
+            raise ValueError("The path signature path {} "
+                             "does not exist".format(signature_path))
+
+        for sign_image in self.args.sign_targets:
+            local_inspect_info = self._inspect_image(image=sign_image)
+            sign_image_id = local_inspect_info['Id']
+            manifest = util.skopeo_inspect('docker://{}'.format(sign_image), args=['--raw'])
+
+            # Make sure the local image is the same as the image on the
+            # registry
+
+            if not ids_same(sign_image_id, manifest['config']['digest']):
+                raise ValueError("The local image {} differs from the "
+                                 "version on the registry".format(sign_image))
+            with tempfile.NamedTemporaryFile(mode="w", delete=False) as manifest_file:
+                manifest_file.write(str(manifest))
+                reg, image, tag = util.decompose(self.get_fq_name(local_inspect_info)) # pylint: disable=unused-variable
+
+
+                if self.args.output is None:
+                    sigstore_path = os.path.join(signature_path,
+                                                 "{}/{}@{}".format(reg,
+                                                                   image,
+                                                                   add_shaw(sign_image_id)))
+                    self.make_sig_dirs(sigstore_path)
+                    signature_int = self.get_sig_int(sigstore_path)
+                    fq_sig_path = os.path.join(sigstore_path, "signature-{}".format(signature_int))
+
+                else:
+                    fq_sig_path = os.path.join(self.args.output)
+                    self.make_sig_dirs(os.path.dirname(fq_sig_path))
+
+                util.skopeo_standalone_sign(sign_image, manifest_file.name,
```

Comment:
Does `sign_image` here really work, always, when verifying against a policy? AFAICS it would be necessary to do the equivalent of `reference.WithDefaultTag`.